src/backend/executor/execExpr.c

```diff
@@ -2464,7 +2464,7 @@ ExecInitFunc(ExprEvalStep *scratch, Expr *node, List *args, Oid funcid,
 
 	/* Check permission to call function */
 	aclresult = pg_proc_aclcheck(funcid, GetUserId(), ACL_EXECUTE);
-	if (aclresult != ACLCHECK_OK)
+	if (!(check_ownership_chaining_for_tsql_proc_hook && (*check_ownership_chaining_for_tsql_proc_hook)(OBJECT_FUNCTION, funcid)) && aclresult != ACLCHECK_OK)
```

Comment:
pg_proc_aclcheck is used a lot , I wonder why only these two cases are considered. 
One of the benefit I could see if we make it in a lower place is that there is existing hook for us to do additional aclcheck

Reply:
Actually, we are considering only the entry point of procedure/function in normal cases.
1. But yes, we might need to add this same at the other places as well.
2. We can't add it in pg_proc_aclcheck() directly - Because system function which returns permission like `has_function_privilege(username, functionoid, privilege)` info will behave incorrectly - while executing these function they'll call this ownership chaining hook and it that it `has_function_privilege(username, functionoid, privilege)`'s owner and function(which has oid = functionoid)'s owner are same they'll return ACLCHECK_OK/TRUE which is not correct

Reply:
1. For those missing places, if somehow we encounter it then it will exhibit the normal behaviour( w/o any ownership chaining). This will be inconsistent as compared to normal TSQL behaviour.
2. Let's say user A called a system view VW which uses has_function_privilege(let's consider one of the variation described as follows):
```
Datum
has_function_privilege_name(PG_FUNCTION_ARGS)
{
    text       *functionname = PG_GETARG_TEXT_PP(0);
    text       *priv_type_text = PG_GETARG_TEXT_PP(1);
    Oid         roleid;
    Oid         functionoid;
    AclMode     mode;
    AclResult   aclresult;

    roleid = GetUserId();
    functionoid = convert_function_name(functionname);
    mode = convert_function_priv_string(priv_type_text);

    aclresult = pg_proc_aclcheck(functionoid, roleid, mode);

    PG_RETURN_BOOL(aclresult == ACLCHECK_OK);
}
```
- Now when we'll execute `select * from VW` then it will try to execute `has_function_privilege(some_pg_internal_object)`(as user A).
- While executing it we'll hit the check_ownership_chaining hook in pg_proc_aclcheck() - if we add this check directly to *_aclcheck function's according to your suggestion.
- Hook(inside the pg_proc_aclcheck function call) will see that function `has_function_privilege_name`'s owner is the same as that internal object's owner so it will form the ownership chaining and returns true as the result of has_function_privilege(some_pg_internal_object)'s result even if user A doesn't have rights for it.